File: 0002-add-two-numbers/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* dummy = new ListNode(-1);
        ListNode* sum_list_ptr = dummy;
        
        ListNode* l1_ptr = l1;
        ListNode* l2_ptr = l2;
        
        int carry = 0;
        while (l1_ptr || l2_ptr) {
            
            int val1 = 0;
            int val2 = 0;
            
            if (l1_ptr != nullptr) {
                val1 = l1_ptr->val;
                l1_ptr = l1_ptr->next;
            }
            
            if (l2_ptr != nullptr) {
                val2 = l2_ptr->val;
                l2_ptr = l2_ptr->next;
            }
            
            int sum = val1 + val2 + carry;
            carry = (sum/10);
            int digit = sum % 10;
            
            sum_list_ptr->next = new ListNode(digit);
            sum_list_ptr = sum_list_ptr->next;
        }
        
        if (carry) {
            sum_list_ptr->next = new ListNode(carry);
        }
        
        return dummy->next;
    
    }
};

File: 0002-add-two-numbers/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) 
    {
        int carry = 0;
        int digit = 0;
        
        ListNode ans = new ListNode();
        ListNode cur = ans;
        
        ListNode head1, head2;
        head1 = l1; head2 = l2;
        
        while(head1 != null || head2 != null)
        {
            digit = carry;
            if(head1 != null)
            {
                digit += head1.val;
                head1 = head1.next;
            }
            if(head2 != null)
            {
                digit += head2.val;
                head2 = head2.next;
            }
            
            if(digit >= 10)
            {
                digit -= 10;
                carry = 1;
            }
            else carry = 0;
            
            cur.next = new ListNode(digit);
            cur      = cur.next;
        }
        
        if(carry == 1) cur.next = new ListNode(carry);
        return ans.next;
        
        
    }
}
